### ISSUE TITLE
TorchSharp RELU does not work in Torch.RL project!

### ISSUE THREAD

**Issue by GeorgeS2019** [dotnet/TorchSharp#446]
Here is the [issue: ](https://github.com/losttech/Torch.RL/issues/1)

```cs
 ActorCritic ActorCriticFactory() {
        // LeakyRELU does not work o-O
        static Module activation() => ReLU(inPlace: false);
}
```

@NiklasGustafsson  => Not sure if this **ReLU issue** has been addressed in a more recent release of TorchSharp.

This project uses "TorchSharp" Version="0.91.52681"

**Comment by NiklasGustafsson** [comment]
A repro would be useful. 

It is not clear to me whether the comment in that code is saying that LeakyReLU() is buggy in TorchSharp, or whether it doesn't work for using in the specific model.

When I augment the LeakyReLU unit test to do a better job testing output values, it gets the right result.

```C#
        [Fact]
        public void EvaluateLeakyRelu()
        {
            var rel = LeakyReLU(0.1);
            var input = torch.randn(new long[] { 64, 8 });
            var output = rel.forward(input);
            var values = output.data<float>().ToArray();
            Assert.Equal(input.shape, output.shape);

            var singleton = torch.tensor(15.0f);
            Assert.Equal(15.0f, rel.forward(singleton).item<float>());
            singleton = torch.tensor(-15.0f);
            Assert.Equal(-1.50f, rel.forward(singleton).item<float>());
        }
```

the interop implementation is nothing but a call to the native code, so I don't know where an error could be introduced:

```C++
NNModule THSNN_LeakyReLU_ctor(const double negative_sloope, const bool inplace, NNAnyModule* outAsAnyModule)
{
    CATCH_RETURN_NNModule(
        auto opts = torch::nn::LeakyReLUOptions().negative_slope(negative_sloope).inplace(inplace);
    res = create_module<torch::nn::LeakyReLUImpl>(opts, outAsAnyModule);
    );
}

Tensor THSNN_LeakyReLU_forward(const NNModule module, const Tensor tensor)
{
    CATCH_TENSOR((*module)->as<torch::nn::LeakyReLU>()->forward(*tensor));
}
```

So, without some specifics on what is going wrong, I have nothing to go on here.

**Comment by GeorgeS2019** [comment]
@NiklasGustafsson  we are trying to evaluate the feasibility to reuse some of the codes in this repos as a proof of concepts of creating a RL template using TorchSharp BUT instead of the usual Unity3D, but a MIT highly popular 3D game engine Godot 3D that support c#.

I have suggest this concept to Luis Quintanilla, He likes it. 

Could you help in your free time to clone the repos and upgrade the repos to the latest TorchSharp => right now the Test codes does not work. If you could upgrade the repo to latest TorchSharp and the Test works, this would be a grreat confidence for us to proceed.

**Comment by NiklasGustafsson** [comment]
@GeorgeS2019, I can certainly clone the repo and test it, if that is the reproduction case. 

However, without a description of how the use of the LeakyReLU activation differs from expected results, I have very little to go on. The current test doesn't even use LeakyReLU, presumably because it doesn't work.

All I have to go on is this:

```C#
                // LeakyRELU does not work o-O
                static Module activation() => ReLU(inPlace: false);
```

Which is not enough to explain in what way "LeakyReLU does not work." I would need some explanation of what behavior is observed relative to what was expected. As I mentioned above, in the augmented unit test, LeakyReLU produces the right result.

**Comment by GeorgeS2019** [comment]
Here is the more complete c# console tensorflow (PythonNET) [codes ](https://github.com/losttech/Gradient-Samples/tree/b8bba77d3a2712bbe66846df77d2726ef570f475/RL-MLAgents)from which ported to TorchSharp RL

Perhaps looking at the Console codes could suggest how to complete the TorchSharp RL code, which is incomplete in the older version of TorchSharp, but no more handicap in the latest TorchSharp.

**Comment by NiklasGustafsson** [comment]
It does not help. 

I will need to have a clear description of what isn't working. That is, observed vs. expected behavior.

**Comment by GeorgeS2019** [comment]
Still not yet there, but here is the [link ](https://github.com/virtualmlnet/hackathon-2021/issues/6)to the idea.

More will be added soon.
